test/App/Fossa/VSI/TypesSpec.hs

```diff
@@ -0,0 +1,141 @@
+import Data.Aeson (eitherDecode)
+import Data.ByteString.Lazy.Char8 qualified as BS
+import Data.Either (isLeft)
+import Data.Map qualified as Map
+import Test.Hspec (Spec, describe, it, shouldBe, shouldMatchList, shouldSatisfy)
+import Test.Hspec.Core.Spec (focus)
+import Text.RawString.QQ
+
+inferencesBody :: BS.ByteString
+inferencesBody =
+  [r|
+{
+  "InferencesByFilePath": {
+    "/foo/bar.h": {
+      "RawSha256": "YmIwMTg2NTNlOTVlY2U5M2VmMDYwMTQ3YjA0ZjZhYzRkZjhlMzFhZDc1OWFjYmExZWJmMjIwZDVjZTJlM2ZkZQ==",
+      "ComponentID": "0f4ba6a8-5b3f-436f-8c36-828e7375aef7",
+      "Locator": "git+github.com/facebook/folly$v2016.08.08.00",
+      "Confidence": 1
+    }
+  }
+}
+|]
+
+emptyLocatorInference :: BS.ByteString
+emptyLocatorInference =
+  [r|
+{
+  "InferencesByFilePath": {
+    "/foo/bar.h": {
+      "RawSha256": "",
+      "ComponentID": "",
+      "Locator": "",
+      "Confidence": 1
+    }
+  }
+}
+|]
+
+invalidLocatorInference :: BS.ByteString
+invalidLocatorInference =
+  [r|
+{
+  "InferencesByFilePath": {
+    "/foo/bar.h": {
+      "RawSha256": "",
+      "ComponentID": "",
+      "Locator": "bad-locator",
+      "Confidence": 1
+    }
+  }
+}
+|]
+
+expectedEmptyLocatorInference :: VsiExportedInferencesBody
+expectedEmptyLocatorInference =
+  VsiExportedInferencesBody $
+    Map.fromList
+      [
+        ( VsiFilePath "/foo/bar.h"
+        , VsiInference Nothing
+        )
+      ]
+
+expectedSingleInference :: VsiExportedInferencesBody
+expectedSingleInference =
+  VsiExportedInferencesBody $ Map.fromList singleInference
+
+singleInference :: [(VsiFilePath, VsiInference)]
+singleInference =
+  [
+    ( VsiFilePath "/foo/bar.h"
+    , VsiInference . Just $ Locator "git" "github.com/facebook/folly" "v2016.08.08.00"
+    )
+  ]
+
+singleRuleExpected :: [VsiRule]
+singleRuleExpected = [VsiRule (VsiRulePath "/foo") (Locator "git" "github.com/facebook/folly" "v2016.08.08.00")]
+
+commonPrefixInferences :: [(VsiFilePath, VsiInference)]
+commonPrefixInferences =
+  (VsiFilePath "/foo/bar/baz.c", VsiInference . Just $ Locator "git" "github.com/facebook/folly" "v2016.08.08.00") : singleInference
+
+multipleInferences :: [(VsiFilePath, VsiInference)]
+multipleInferences =
+  (VsiFilePath "/otherProject/Readme.md", VsiInference . Just $ Locator "git" "github.com/otherProject" "2.0.0")
+    : (VsiFilePath "/baz/hello.c", VsiInference . Just $ Locator "git" "github.com/someProject" "1.0.0")
+    : commonPrefixInferences
+
+multipleRulesExpected :: [VsiRule]
+multipleRulesExpected =
+  VsiRule (VsiRulePath "/otherProject") (Locator "git" "github.com/otherProject" "2.0.0")
+    : VsiRule (VsiRulePath "/baz") (Locator "git" "github.com/someProject" "1.0.0")
+    : singleRuleExpected
+
+nestedProjectInferences :: [(VsiFilePath, VsiInference)]
+nestedProjectInferences =
+  ( VsiFilePath "/foo/bar/g.c"
+  , VsiInference . Just $ Locator "git" "github.com/facebook/follyNested" "1.0.0"
+  )
+    : singleInference
```

Comment:
Technically this cannot occur with the current design of VSI, but since it's written it doesn't hurt to support this case!